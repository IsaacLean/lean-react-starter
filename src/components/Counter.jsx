// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';

type Props = {
  onDecrement: Function,
  onIncrement: Function,
  name?: string,
  value: number
};

class Counter extends Component<Props> {
  constructor(props: Props) {
    super(props);

    // this.increment = this.increment.bind(this);
    // this.incrementAsync = this.incrementAsync.bind(this);
    this.incrementIfOdd = this.incrementIfOdd.bind(this);
    // this.decrement = this.decrement.bind(this);
  }

  incrementIfOdd: Function;

  incrementIfOdd() {
    if (this.props.value % 2 !== 0) {
      this.props.onIncrement();
    }
  }

  render() {
    const { onDecrement, onIncrement, value } = this.props;
    return (
      <p>
        <b>{this.props.name}</b>
        <br /> Clicked: {value} times<br />
        <button onClick={onIncrement}>+</button>
        <button onClick={onDecrement}>-</button>
        <button onClick={this.incrementIfOdd}>Increment if odd</button>
      </p>
    );
  }
}

const mapStateToProps = state => ({
  value: state
});
const mapDispatchToProps = dispatch => ({
  onDecrement() {
    dispatch({ type: 'DECREMENT' });
  },

  onIncrement() {
    dispatch({ type: 'INCREMENT' });
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Counter);
