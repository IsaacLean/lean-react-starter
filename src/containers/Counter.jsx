// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';

type Props = {
  onDecrement: Function,
  onIncrement: Function,
  name?: string,
  value: number
};

class Counter extends Component<Props> {
  constructor(props: Props) {
    super(props);

    this.incrementAsync = this.incrementAsync.bind(this);
    this.incrementIfOdd = this.incrementIfOdd.bind(this);
  }

  incrementAsync: Function;
  incrementAsync() {
    setTimeout(this.props.onIncrement, 1000);
  }

  incrementIfOdd: Function;
  incrementIfOdd() {
    if (this.props.value % 2 !== 0) {
      this.props.onIncrement();
    }
  }

  render() {
    const { onDecrement, onIncrement, value } = this.props;
    return (
      <div>
        <b>{this.props.name}</b>
        <p>Clicked: {value} times</p>
        <button onClick={onIncrement}>+</button>
        <button onClick={onDecrement}>-</button>
        <button onClick={this.incrementIfOdd}>Increment if odd</button>
        <button onClick={this.incrementAsync}>Increment async</button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  value: state
});
const mapDispatchToProps = dispatch => ({
  onDecrement() {
    dispatch({ type: 'DECREMENT' });
  },

  onIncrement() {
    dispatch({ type: 'INCREMENT' });
  }
});

export const Unwrapped = Counter;
export default connect(mapStateToProps, mapDispatchToProps)(Counter);
