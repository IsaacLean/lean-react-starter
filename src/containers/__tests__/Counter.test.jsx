import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import React from 'react';

import { Unwrapped as UnwrappedCounter } from '../Counter';

Enzyme.configure({ adapter: new Adapter() });

function setup(value = 0) {
  const actions = {
    onDecrement: jest.fn(),
    onIncrement: jest.fn()
  };

  const component = Enzyme.shallow(<UnwrappedCounter value={value} {...actions} />);

  return {
    actions,
    buttons: component.find('button'),
    component,
    p: component.find('p')
  };
}

describe('Counter component', () => {
  it('renders correctly', () => {
    const { component } = setup();

    expect(component).toMatchSnapshot();
  });

  it('should display count', () => {
    const { p } = setup();
    expect(p.text()).toMatch(/^Clicked: 0 times/);
  });

  it('first button should call onIncrement', () => {
    const { actions, buttons } = setup();
    buttons.at(0).simulate('click');
    expect(actions.onIncrement).toBeCalled();
  });

  it('second button should call onDecrement', () => {
    const { actions, buttons } = setup();
    buttons.at(1).simulate('click');
    expect(actions.onDecrement).toBeCalled();
  });

  it('third button should not call onIncrement if the counter is even', () => {
    const { actions, buttons } = setup(42);
    buttons.at(2).simulate('click');
    expect(actions.onIncrement).not.toBeCalled();
  });

  it('third button should call onIncrement if the counter is odd', () => {
    const { actions, buttons } = setup(43);
    buttons.at(2).simulate('click');
    expect(actions.onIncrement).toBeCalled();
  });

  it('third button should call onIncrement if the counter is odd', () => {
    const { actions, buttons } = setup(-43);
    buttons.at(2).simulate('click');
    expect(actions.onIncrement).toBeCalled();
  });

  it('fourth button should call onCremenet in a second', done => {
    const { actions, buttons } = setup();
    buttons.at(3).simulate('click');
    setTimeout(() => {
      expect(actions.onIncrement).toBeCalled();
      done();
    }, 1000);
  });
});
