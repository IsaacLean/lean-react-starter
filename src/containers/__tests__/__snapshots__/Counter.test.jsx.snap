// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component renders correctly 1`] = `
<div>
  <b />
  <p>
    Clicked: 
    0
     times
  </p>
  <button
    onClick={[Function]}
  >
    +
  </button>
  <button
    onClick={[Function]}
  >
    -
  </button>
  <button
    onClick={[Function]}
  >
    Increment if odd
  </button>
  <button
    onClick={[Function]}
  >
    Increment async
  </button>
</div>
`;
